### PR TITLE
Towards a dictionary builder

### PR DESCRIPTION
**Pull request by Yoric**

### CODE REVIEW

**Comment by arai-a** [comment]
can you summarize the goal here?

**Comment by arai-a** [comment]
also, the fix for #176 seems to be mixed into mutliple PRs
what's happening?

**Comment by Yoric** [comment]
> also, the fix for #176 seems to be mixed into mutliple PRs
what's happening?

Well, it's a complicated refactoring. I added new features in one PR, refactored the code that used the old features in another PR.

**Comment by Yoric** [comment]
I should also clarify that the format used to store dictionaries is temporary. Actual format is being discussed [here](https://github.com/binast/container-format-spec/issues/7).

File: src/bin/generate_dictionary.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                .takes_value(true)
+                .default_value("0")
+                .validator(|s| s.parse::<u32>()
+                    .map(|_| ())
+                    .map_err(|e| format!("Invalid number {}", e)))
+                .help("Number of layers of functions to lazify. 0 = no lazification, 1 = functions at toplevel, 2 = also functions in functions at toplevel, etc."),
+            Arg::with_name("quiet")
+                .long("quiet")
+                .short("q")
+                .help("Do not print progress"),
```
[the remark this answers is not in the document]
Comment (Yoric) [reply src/bin/generate_dictionary.rs]: Yes, we already did some of that with the `advanced` subcommand, but we'll probably want to go further.

File: crates/binjs_io/src/io/deprecated.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        self.string(string)
+    }
+
+    fn identifier_name(&mut self, value: Option<&IdentifierName>) -> Result<Self::Tree, TokenWriterError> {
+        let string = value.map(IdentifierName::as_shared_string);
+        self.string(string)
+    }
+}
+
+pub struct TokenWriterTreeAdapter<T> where T: TokenWriterWithTree {
```
[the remark this answers is not in the document]
Comment (Yoric) [reply crates/binjs_io/src/io/deprecated.rs]: Good idea.
